File: python/src/onchaindiligence/agent_evidence/canonical.py
"""Strict JSON parsing and RFC 8785 canonicalization."""

from __future__ import annotations

import json
import math
import re
from datetime import datetime, timezone
from hashlib import sha256
from typing import Any

import rfc8785

from .errors import CanonicalizationError, ParseError
from .models import JsonObject

_TIMESTAMP = re.compile(r"^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}Z$")
_MAX_SAFE_INTEGER = 9_007_199_254_740_991


def _object_no_duplicates(pairs: list[tuple[str, Any]]) -> JsonObject:
    result: JsonObject = {}
    for key, value in pairs:
        if key in result:
            raise ParseError(f"duplicate JSON object key: {key}")
        result[key] = value
    return result


def _safe_integer(value: str) -> int:
    parsed = int(value)
    if abs(parsed) > _MAX_SAFE_INTEGER:
        raise ParseError("JSON integer exceeds the interoperable safe-integer range")
    return parsed


def _finite_float(value: str) -> float:
    parsed = float(value)
    if not math.isfinite(parsed):
        raise ParseError("JSON number exceeds the finite IEEE-754 range")
    return parsed


def _reject_constant(value: str) -> None:
    raise ParseError(f"non-finite JSON number is forbidden: {value}")


def parse_json(data: str | bytes) -> Any:
    """Parse untrusted JSON with duplicate-name and unsafe-integer rejection."""

    if isinstance(data, bytes):
        try:
            text = data.decode("utf-8", errors="strict")
        except UnicodeDecodeError as exc:
            raise ParseError("JSON must be valid UTF-8") from exc
    else:
        text = data
    try:
        return json.loads(
            text,
            object_pairs_hook=_object_no_duplicates,
            parse_int=_safe_integer,
            parse_float=_finite_float,
            parse_constant=_reject_constant,
        )
    except ParseError:
        raise
    except (json.JSONDecodeError, UnicodeError, ValueError, RecursionError) as exc:
        raise ParseError(f"invalid JSON: {exc}") from exc


def enforce_limits(
    value: Any,
    *,
    max_depth: int,
    max_string_length: int,
    max_array_length: int,
) -> None:
    """Enforce policy limits not expressible across every open schema value."""

    def visit(item: Any, depth: int) -> None:
        if depth > max_depth:
            raise ParseError(f"JSON exceeds maximum depth {max_depth}")
        if isinstance(item, str):
            if len(item) > max_string_length:
                raise ParseError(f"JSON string exceeds maximum length {max_string_length}")
            return
        if isinstance(item, list):
            if len(item) > max_array_length:
                raise ParseError(f"JSON array exceeds maximum length {max_array_length}")
            for child in item:
                visit(child, depth + 1)
            return
        if isinstance(item, dict):
            if len(item) > max_array_length:
                raise ParseError(f"JSON object exceeds maximum members {max_array_length}")
            for key, child in item.items():
                if not isinstance(key, str):
                    raise ParseError("JSON object names must be strings")
                visit(key, depth + 1)
                visit(child, depth + 1)
            return
        if isinstance(item, int) and not isinstance(item, bool) and abs(item) > _MAX_SAFE_INTEGER:
            raise ParseError("JSON integer exceeds the interoperable safe-integer range")
        if isinstance(item, float) and (item != item or item in (float("inf"), float("-inf"))):
            raise ParseError("non-finite JSON number is forbidden")

    visit(value, 0)


def canonicalize(value: Any) -> bytes:
    """Return RFC 8785 bytes or raise a stable package exception."""

    try:
        return rfc8785.dumps(value)
    except (rfc8785.CanonicalizationError, TypeError, ValueError) as exc:
        raise CanonicalizationError(str(exc)) from exc


def content_id(value: Any) -> str:
    """Compute the v0 SHA-256 unpadded-base64url content identifier."""

    import base64

    digest = base64.urlsafe_b64encode(sha256(canonicalize(value)).digest()).decode("ascii").rstrip("=")
    return f"sha256:{digest}"


def parse_timestamp(value: str) -> datetime:
    """Parse the exact millisecond UTC timestamp profile."""

    if not _TIMESTAMP.fullmatch(value):
        raise ParseError("timestamp must use exact YYYY-MM-DDTHH:mm:ss.sssZ syntax")
    try:
        parsed = datetime.fromisoformat(value[:-1] + "+00:00")
    except ValueError as exc:
        raise ParseError(f"invalid UTC timestamp: {value}") from exc
    return parsed.astimezone(timezone.utc)


def format_timestamp(value: datetime) -> str:
    """Format an aware datetime in the normative millisecond UTC profile."""

    if value.tzinfo is None:
        raise ValueError("timestamp datetime must be timezone-aware")
    return value.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")

File: packages/agent-evidence/src/canonical.ts
import { createHash } from 'node:crypto'
import { CanonicalizationError, ParseError } from './errors.js'
import type { JsonValue } from './types.js'

const TIMESTAMP = /^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}Z$/

function assertUnicodeScalarValue(value: string): void {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index)
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1)
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new CanonicalizationError('canonical JSON cannot contain an unpaired surrogate')
      }
      index += 1
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      throw new CanonicalizationError('canonical JSON cannot contain an unpaired surrogate')
    }
  }
}

function canonical(value: unknown, seen: Set<object>): string {
  if (value === null || typeof value === 'boolean') return JSON.stringify(value)
  if (typeof value === 'string') {
    assertUnicodeScalarValue(value)
    return JSON.stringify(value)
  }
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) throw new CanonicalizationError('canonical JSON cannot contain non-finite numbers')
    if (Number.isInteger(value) && !Number.isSafeInteger(value) && !JSON.stringify(value).includes('e')) {
      throw new CanonicalizationError('canonical JSON integer exceeds the interoperable safe-integer range')
    }
    return JSON.stringify(value)
  }
  if (typeof value !== 'object') {
    throw new CanonicalizationError(`value of type ${typeof value} is not valid JSON`)
  }
  if (seen.has(value)) throw new CanonicalizationError('canonical JSON cannot contain cycles')
  seen.add(value)
  try {
    if (Array.isArray(value)) return `[${value.map((item) => canonical(item, seen)).join(',')}]`
    const prototype = Object.getPrototypeOf(value)
    if (prototype !== Object.prototype && prototype !== null) {
      throw new CanonicalizationError('canonical JSON objects must have a plain object prototype')
    }
    const record = value as Record<string, unknown>
    const entries = Object.keys(record)
      .sort()
      .map((key) => {
        assertUnicodeScalarValue(key)
        return `${JSON.stringify(key)}:${canonical(record[key], seen)}`
      })
    return `{${entries.join(',')}}`
  } finally {
    seen.delete(value)
  }
}

export function canonicalize(value: unknown): Uint8Array {
  return Buffer.from(canonical(value, new Set()), 'utf8')
}

export function canonicalizeText(value: unknown): string {
  return Buffer.from(canonicalize(value)).toString('utf8')
}

export function contentId(value: unknown): string {
  return `sha256:${createHash('sha256').update(canonicalize(value)).digest('base64url')}`
}

export function formatTimestamp(value: Date): string {
  if (!(value instanceof Date) || Number.isNaN(value.getTime())) throw new TypeError('timestamp must be a valid Date')
  return value.toISOString()
}

export function parseTimestamp(value: string): Date {
  if (!TIMESTAMP.test(value)) throw new ParseError('timestamp must use exact YYYY-MM-DDTHH:mm:ss.sssZ syntax')
  const parsed = new Date(value)
  if (Number.isNaN(parsed.getTime()) || parsed.toISOString() !== value) {
    throw new ParseError(`invalid UTC timestamp: ${value}`)
  }
  return parsed
}

export function cloneJson<T extends JsonValue>(value: T): T {
  return parseJson(canonicalize(value)) as T
}

export function enforceLimits(
  value: unknown,
  limits: { maxDepth: number; maxStringLength: number; maxArrayLength: number },
): void {
  const visit = (item: unknown, depth: number): void => {
    if (depth > limits.maxDepth) throw new ParseError(`JSON exceeds maximum depth ${limits.maxDepth}`)
    if (typeof item === 'string') {
      if (item.length > limits.maxStringLength) {
        throw new ParseError(`JSON string exceeds maximum length ${limits.maxStringLength}`)
      }
      return
    }
    if (Array.isArray(item)) {
      if (item.length > limits.maxArrayLength) {
        throw new ParseError(`JSON array exceeds maximum length ${limits.maxArrayLength}`)
      }
      for (const child of item) visit(child, depth + 1)
      return
    }
    if (item !== null && typeof item === 'object') {
      const entries = Object.entries(item)
      if (entries.length > limits.maxArrayLength) {
        throw new ParseError(`JSON object exceeds maximum members ${limits.maxArrayLength}`)
      }
      for (const [key, child] of entries) {
        visit(key, depth + 1)
        visit(child, depth + 1)
      }
      return
    }
    if (typeof item === 'number') {
      if (!Number.isFinite(item)) throw new ParseError('non-finite JSON number is forbidden')
      if (Number.isInteger(item) && !Number.isSafeInteger(item) && !JSON.stringify(item).includes('e')) {
        throw new ParseError('JSON integer exceeds the interoperable safe-integer range')
      }
    }
  }
  visit(value, 0)
}

export function parseJson(data: string | Uint8Array): JsonValue {
  const text = typeof data === 'string' ? data : new TextDecoder('utf-8', { fatal: true }).decode(data)
  scanStrictJson(text)
  try {
    return JSON.parse(text) as JsonValue
  } catch (error) {
    throw new ParseError(`invalid JSON: ${error instanceof Error ? error.message : String(error)}`, { cause: error })
  }
}

function scanStrictJson(text: string): void {
  let index = 0
  const whitespace = (): void => {
    while (text[index] === ' ' || text[index] === '\t' || text[index] === '\r' || text[index] === '\n') index += 1
  }
  const stringToken = (): string => {
    const start = index
    index += 1
    while (index < text.length) {
      const character = text[index]
      if (character === '\\') {
        index += 1
        const escaped = text[index]
        if (escaped === 'u') {
          if (!/^[0-9a-fA-F]{4}$/.test(text.slice(index + 1, index + 5))) throw new ParseError('invalid JSON unicode escape')
          index += 5
        } else if ('"\\/bfnrt'.includes(escaped ?? '')) index += 1
        else throw new ParseError('invalid JSON escape')
      } else if (character === '"') {
        index += 1
        const raw = text.slice(start, index)
        const parsed = JSON.parse(raw) as string
        assertUnicodeScalarValue(parsed)
        return parsed
      } else {
        if (character === undefined || character.charCodeAt(0) < 0x20) throw new ParseError('invalid JSON string')
        index += 1
      }
    }
    throw new ParseError('unterminated JSON string')
  }
  const value = (depth: number): void => {
    if (depth > 64) throw new ParseError('JSON exceeds maximum parser depth 64')
    whitespace()
    const character = text[index]
    if (character === '{') {
      index += 1
      whitespace()
      const keys = new Set<string>()
      if (text[index] === '}') { index += 1; return }
      while (true) {
        whitespace()
        if (text[index] !== '"') throw new ParseError('expected JSON object key')
        const key = stringToken()
        if (keys.has(key)) throw new ParseError(`duplicate JSON object key: ${key}`)
        keys.add(key)
        whitespace()
        if (text[index] !== ':') throw new ParseError('expected colon')
        index += 1
        value(depth + 1)
        whitespace()
        if (text[index] === '}') { index += 1; return }
        if (text[index] !== ',') throw new ParseError('expected comma')
        index += 1
      }
    }
    if (character === '[') {
      index += 1
      whitespace()
      if (text[index] === ']') { index += 1; return }
      while (true) {
        value(depth + 1)
        whitespace()
        if (text[index] === ']') { index += 1; return }
        if (text[index] !== ',') throw new ParseError('expected comma')
        index += 1
      }
    }
    if (character === '"') { stringToken(); return }
    const token = text.slice(index).match(/^(?:true|false|null|-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?)/)?.[0]
    if (!token) throw new ParseError('invalid JSON value')
    if (/^-?(?:0|[1-9]\d*)$/.test(token) && !Number.isSafeInteger(Number(token))) {
      throw new ParseError('JSON integer exceeds the interoperable safe-integer range')
    }
    if (/^-?(?:0|[1-9]\d*)(?:\.\d+|[eE])/.test(token) && !Number.isFinite(Number(token))) {
      throw new ParseError('JSON number exceeds the finite IEEE-754 range')
    }
    index += token.length
  }
  value(0)
  whitespace()
  if (index !== text.length) throw new ParseError('unexpected trailing JSON data')
}
